daemon/list.go

```diff
@@ -103,55 +106,72 @@ func (daemon *Daemon) Containers(ctx context.Context, config *containertypes.Lis
 		return nil, err
 	}
 
-	var (
-		view       = daemon.containersReplica.Snapshot()
-		containers = []*containertypes.Summary{}
-	)
-
+	view := daemon.containersReplica.Snapshot()
 	filter, err := daemon.foldFilter(ctx, view, config)
 	if err != nil {
 		return nil, err
 	}
 
-	// fastpath to only look at a subset of containers if specific name
+	// shortcut to only look at a subset of containers if specific name
 	// or ID matches were provided by the user--otherwise we potentially
 	// end up querying many more containers than intended
+	//
+	// TODO (thaJeztah): given containersReplica.Snapshot() provides a "consistent read-only view of the database" (which indicates "de-referenced copy", is there any reason we wouldn't use a []*container.Snapshot (pointer-slice)?.
 	containerList, err := daemon.filterByNameIDMatches(view, filter)
 	if err != nil {
 		return nil, err
 	}
+	numContainers := len(containerList)
+
+	// Get the info for each container in the list; this can be slow so we
+	// dispatch a set number of worker goroutines to do the jobs. We choose
+	// log2(numContainers) workers to avoid creating too many goroutines
+	// for large number of containers.
+	numWorkers := int(math.Log2(float64(numContainers)))
+	if numWorkers < 1 {
+		numWorkers = 1
+	}
 
-	for i := range containerList {
-		currentContainer := &containerList[i]
-		switch includeContainerInList(currentContainer, filter) {
+	resultsMut := sync.Mutex{}
+	results := make([]*containertypes.Summary, 0, min(len(containerList), filter.Limit))
+	g, ctx := errgroup.WithContext(ctx)
+	g.SetLimit(numWorkers)
+
+	for _, ctr := range containerList {
+		switch includeContainerInList(&ctr, filter) {
 		case excludeContainer:
 			continue
 		case stopIteration:
-			return containers, nil
-		}
+			break
+		default:
+			filter.idx++
 
-		// transform internal container struct into api structs
-		newC, err := daemon.refreshImage(ctx, currentContainer)
-		if err != nil {
-			return nil, err
+			g.Go(func() error {
+				// refresh the container image info (in case the image changed in
+				// the repository)
+				newC := daemon.refreshImage(ctx, &ctr)
+
+				// get the image size (calculation is slow)
+				if filter.Size {
+					var err error
+					newC.SizeRw, newC.SizeRootFs, err = daemon.imageService.GetContainerLayerSize(ctx, newC.ID)
+					if err != nil {
+						return err
+					}
+				}
+				resultsMut.Lock()
+				results = append(results, newC)
+				resultsMut.Unlock()
+				return nil
+			})
 		}
+	}
 
-		// release lock because size calculation is slow
-		if filter.Size {
-			sizeRw, sizeRootFs, err := daemon.imageService.GetContainerLayerSize(ctx, newC.ID)
-			if err != nil {
-				return nil, err
-			}
-			newC.SizeRw = sizeRw
-			newC.SizeRootFs = sizeRootFs
-		}
-		if newC != nil {
-			containers = append(containers, newC)
-			filter.idx++
-		}
+	if err := g.Wait(); err != nil {
+		return nil, err
 	}
 
-	return containers, nil
+	return results, nil
```

Comment:
```suggestion
	return results[:filter.idx], nil
```

Reply:
Makes sense, but should be `filter.idx` (not `filter.idx+1`) because of the earlier `filter.idx++`. The unit tests confirm this.